cpp/src/strings/regex/regexec.cu

```diff
@@ -120,32 +115,53 @@ std::unique_ptr<reprog_device, std::function<void(reprog_device*)>> reprog_devic
   }
 
   // initialize the rest of the elements
-  d_prog->_codepoint_flags = codepoint_flags;
-
-  // allocate execute memory if needed
-  rmm::device_buffer* d_relists{};
-  if (insts_count > RX_LARGE_INSTS) {
-    // two relist state structures are needed for execute per string
-    auto const rlm_size  = relist::alloc_size(insts_count) * 2 * strings_count;
-    d_relists            = new rmm::device_buffer(rlm_size, stream);
-    d_prog->_relists_mem = d_relists->data();
-  }
+  d_prog->_max_insts = insts_count;
+  d_prog->_prog_size = memsize + sizeof(reprog_device);
 
   // copy flat prog to device memory
   CUDF_CUDA_TRY(cudaMemcpyAsync(
     d_buffer->data(), h_buffer.data(), memsize, cudaMemcpyHostToDevice, stream.value()));
 
   // build deleter to cleanup device memory
-  auto deleter = [d_buffer, d_relists](reprog_device* t) {
+  auto deleter = [d_buffer](reprog_device* t) {
     t->destroy();
     delete d_buffer;
-    delete d_relists;
   };
+
   return std::unique_ptr<reprog_device, std::function<void(reprog_device*)>>(d_prog, deleter);
 }
 
 void reprog_device::destroy() { delete this; }
 
+std::size_t reprog_device::working_memory_size(int32_t num_threads) const
+{
+  return relist::alloc_size(_insts_count, num_threads) * 2;
+}
+
+std::pair<std::size_t, int32_t> reprog_device::compute_strided_working_memory(
+  int32_t rows, int32_t min_rows, std::size_t max_size) const
+{
+  auto thread_count = rows;
+  auto buffer_size  = working_memory_size(thread_count);
+  while ((buffer_size > max_size) && (thread_count > min_rows)) {
+    thread_count = thread_count / 2;
+    buffer_size  = working_memory_size(thread_count);
+  }
+  return std::make_pair(buffer_size, thread_count);
```

Comment:
Do you need to ever worry about the case where `buffer_size` is still > `max_size` after this while? If `thread_count` drops below `min_rows` you could leave the while with a `buffer_size` that is too large. I assume the converse can't be true as the buffer size required depends on the number of threads and if there are too many threads the buffer size will be large. Also, `thread_count` doesn't get forced back to `min_rows`. Is that variable named well or is it an issue that `thread_count` could be below `min_rows`?

Reply:
It is ok if the `buffer_size` is greater than `max_size` if we hit the `min_rows`. I see what you mean about the `thread_count` going below `min_rows`. The intention was the `min_rows` restriction would override the `max_size` one.

Reply:
I've updated the logic to clamp to `min_rows` and update the doxygen to explain the behavior. Let me know what you think. The intention is to expose this through an API in a later PR to help callers predict memory usage for given regex pattern and column size.

Reply:
This makes sense to me and I like the changes you made, thank you for that.

Is `max_size` a good name here if it isn't the max size. Maybe `preferred_size` or `desired_max`?

Reply:
I'll change it to `requested_max_size`.

Reply:
Thanks for humoring my pedantic requests. 🥇